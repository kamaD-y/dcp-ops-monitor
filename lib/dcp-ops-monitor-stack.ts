import * as path from 'node:path';
import { PythonFunction } from '@aws-cdk/aws-lambda-python-alpha';
import * as cdk from 'aws-cdk-lib';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as cw_actions from 'aws-cdk-lib/aws-cloudwatch-actions';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as sns_subs from 'aws-cdk-lib/aws-sns-subscriptions';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import type { Construct } from 'constructs';

export interface DcpOpsMonitorStackProps extends cdk.StackProps {
  logLevel: string;
  loginUrl: string;
  userAgent: string;
  lineMessageApiUrl: string;
  lineMessageApiToken: string;
}

export class DcpOpsMonitorStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: DcpOpsMonitorStackProps) {
    super(scope, id, props);

    // Parameter Store
    const loginParametersForScraping = ssm.StringParameter.fromSecureStringParameterAttributes(
      this,
      'LoginParametersForScraping',
      {
        parameterName: '/dcp-ops-monitor/login-parameters-for-scraping',
      },
    );

    // SNS Topic
    const successTopic = new sns.Topic(this, 'SuccessTopic', {});
    const failureTopic = new sns.Topic(this, 'FailureTopic', {});

    // S3 Bucket
    const errorBucket = new s3.Bucket(this, 'ErrorBucket', {
      versioned: false,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Lambda Function用の LogGroup
    const logGroup = new logs.LogGroup(this, 'LogGroup', {
      retention: logs.RetentionDays.ONE_WEEK,
    });

    // Lambda Function
    const etlFunction = new lambda.DockerImageFunction(this, 'ETLFunction', {
      code: lambda.DockerImageCode.fromImageAsset(path.join(__dirname, '../lambda/dcp_etl'), {
        file: 'Dockerfile',
        extraHash: props.env?.region,
      }),
      memorySize: 1024,
      timeout: cdk.Duration.seconds(60),
      loggingFormat: lambda.LoggingFormat.JSON,
      logGroup: logGroup,
      applicationLogLevel: props.logLevel,
      environment: {
        POWERTOOLS_LOG_LEVEL: props.logLevel,
        LOGIN_URL: props.loginUrl,
        USER_AGENT: props.userAgent,
        LOGIN_PARAMETER_ARN: loginParametersForScraping.parameterArn,
        SNS_TOPIC_ARN: successTopic.topicArn,
        ERROR_BUCKET_NAME: errorBucket.bucketName,
      },
    });
    etlFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['ssm:GetParameter'],
        resources: [loginParametersForScraping.parameterArn],
      }),
    );
    etlFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['sns:Publish'],
        resources: [successTopic.topicArn],
      }),
    );
    etlFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['s3:PutObject'],
        resources: [`${errorBucket.bucketArn}/*`],
      }),
    );

    // LINE通知用Lambda Function
    const notificationFunction = new PythonFunction(this, 'NotificationFunction', {
      runtime: lambda.Runtime.PYTHON_3_13,
      entry: path.join(__dirname, '../lambda/dcp_notification'),
      index: 'src/handler.py',
      handler: 'handler',
      bundling: {
        assetExcludes: ['.venv'],
      },
      environment: {
        POWERTOOLS_LOG_LEVEL: props.logLevel,
        LINE_MESSAGE_API_URL: props.lineMessageApiUrl,
        LINE_MESSAGE_API_TOKEN: props.lineMessageApiToken,
      },
    });

    notificationFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['logs:DescribeMetricFilters'],
        resources: ['*'],
      }),
    );

    notificationFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['logs:FilterLogEvents'],
        resources: [logGroup.logGroupArn],
      }),
    );

    // SNS Topic に LINE通知用Lambda Function をサブスクライブ
    successTopic.addSubscription(new sns_subs.LambdaSubscription(notificationFunction));
    failureTopic.addSubscription(new sns_subs.LambdaSubscription(notificationFunction));

    // 毎週月曜日 09:00 に実行する Rule を作成
    new events.Rule(this, 'EventRule', {
      schedule: events.Schedule.cron({
        minute: '0',
        hour: '0',
        weekDay: 'MON',
      }),
      targets: [new targets.LambdaFunction(etlFunction)],
    });

    // Lambda Functionエラーログ検知用の MetricFilter を作成
    const etlFunctionMetricFilter = etlFunction.logGroup.addMetricFilter('ETLFunctionErrorLogFilter', {
      filterPattern: logs.FilterPattern.all(logs.FilterPattern.stringValue('$.level', '=', 'ERROR')),
      metricNamespace: 'Custom/DcpOpsMonitor',
      metricName: 'ETLFunctionErrorLog',
      metricValue: '1',
    });

    // MetricFilter に対する CloudWatchAlarm を作成
    etlFunctionMetricFilter
      .metric({
        statistic: cloudwatch.Stats.SUM,
      })
      .createAlarm(this, 'ETLFunctionErrorLogAlarm', {
        threshold: 1,
        evaluationPeriods: 1,
        alarmName: 'ETLFunctionErrorLogAlarm',
        comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      })
      .addAlarmAction(new cw_actions.SnsAction(failureTopic));
  }
}
