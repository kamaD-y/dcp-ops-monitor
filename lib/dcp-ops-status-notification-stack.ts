import * as cdk from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as cw_actions from "aws-cdk-lib/aws-cloudwatch-actions";
import * as logs from "aws-cdk-lib/aws-logs";
import * as iam from "aws-cdk-lib/aws-iam";
import * as events from "aws-cdk-lib/aws-events";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as sns from "aws-cdk-lib/aws-sns";
import * as sns_subs from "aws-cdk-lib/aws-sns-subscriptions";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as targets from "aws-cdk-lib/aws-events-targets";
import { Construct } from "constructs";
import { PythonFunction } from "@aws-cdk/aws-lambda-python-alpha";
import * as path from "path";

export interface DcpOpsMonitorStackProps extends cdk.StackProps {
  logLevel: string;
  loginUrl: string;
  userId: string;
  password: string;
  birthdate: string;
  userAgent: string;
}

export class DcpOpsMonitorStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: DcpOpsMonitorStackProps) {
    super(scope, id, props);

    // Parameter Store
    const loginParametersForScraping = new ssm.StringParameter(
      this,
      "LoginParametersForScraping",
      {
        parameterName: "/custom/dcp_etl/login-parameters",
        stringValue: JSON.stringify({
          USER_ID: props.userId,
          PASSWORD: props.password,
          BIRTHDATE: props.birthdate,
        }),
      }
    );

    // SNS Topic
    const successTopic = new sns.Topic(this, "SuccessTopic", {});
    const failureTopic = new sns.Topic(this, "FailureTopic", {});

    // S3 Bucket
    const errorBucket = new s3.Bucket(this, "ErrorBucket", {
      versioned: false,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    // Lambda Function用の LogGroup
    const logGroup = new logs.LogGroup(this, "LogGroup", {
      retention: logs.RetentionDays.ONE_WEEK,
    });

    // Lambda Function
    const etlFunction = new lambda.DockerImageFunction(this, "EtlFunction", {
      code: lambda.DockerImageCode.fromImageAsset(
        path.join(__dirname, "../lambda/dcp_etl"),
        {
          file: "Dockerfile",
          extraHash: props.env!.region,
        }
      ),
      memorySize: 1024,
      timeout: cdk.Duration.seconds(60),
      loggingFormat: lambda.LoggingFormat.JSON,
      logGroup: logGroup,
      applicationLogLevel: props.logLevel,
      environment: {
        POWERTOOLS_LOG_LEVEL: props.logLevel,
        LOGIN_URL: props.loginUrl,
        USER_AGENT: props.userAgent,
        LOGIN_PARAMETER_ARN: loginParametersForScraping.parameterArn,
        SNS_TOPIC_ARN: successTopic.topicArn,
        ERROR_BUCKET_NAME: errorBucket.bucketName,
      },
    });
    etlFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["ssm:GetParameter"],
        resources: [loginParametersForScraping.parameterArn],
      })
    );
    etlFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["sns:Publish"],
        resources: [successTopic.topicArn],
      })
    );

    // LINE通知用Lambda Function
    const notificationFunction = new PythonFunction(
      this,
      "NotificationFunction",
      {
        runtime: lambda.Runtime.PYTHON_3_13,
        entry: path.join(__dirname, "../lambda/notification"),
        index: "src/handler.py",
        handler: "handler",
        bundling: {
          assetExcludes: [".venv"],
        },
      }
    );

    // SNS Topic に LINE通知用Lambda Function をサブスクライブ
    successTopic.addSubscription(
      new sns_subs.LambdaSubscription(notificationFunction)
    );
    failureTopic.addSubscription(
      new sns_subs.LambdaSubscription(notificationFunction)
    );

    // 毎週月曜日 09:00 に実行する Rule を作成
    new events.Rule(this, "ScrapingNrkExecuteRule", {
      schedule: events.Schedule.cron({
        minute: "0",
        hour: "0",
        weekDay: "MON",
      }),
      targets: [new targets.LambdaFunction(etlFunction)],
    });

    // Lambda Functionエラーログ検知用の MetricFilter を作成
    const etlFunctionMetricFilter = etlFunction.logGroup.addMetricFilter(
      "ETLFunctionErrorLogFilter",
      {
        filterPattern: logs.FilterPattern.all(
          logs.FilterPattern.stringValue("$.level", "=", "ERROR")
        ),
        metricNamespace: "Custom/DcpOpsMonitor",
        metricName: "ETLFunctionErrorLog",
        metricValue: "1",
      }
    );

    // MetricFilter に対する CloudWatchAlarm を作成
    etlFunctionMetricFilter
      .metric({
        statistic: cloudwatch.Stats.SUM,
      })
      .createAlarm(this, "ETLFunctionErrorLogAlarm", {
        threshold: 1,
        evaluationPeriods: 1,
        alarmName: "ETLFunctionErrorLogAlarm",
        comparisonOperator:
          cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      })
      .addAlarmAction(new cw_actions.SnsAction(failureTopic));
  }
}
