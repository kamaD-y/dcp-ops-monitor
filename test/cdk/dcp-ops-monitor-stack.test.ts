import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DcpOpsMonitorStack } from '../../lib/dcp-ops-monitor-stack';

test('Snapshot test for DcpOpsStatusNotificationStack', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new DcpOpsMonitorStack(app, 'DcpOpsMonitorStack', {
    env: {
      account: process.env.CDK_DEFAULT_ACCOUNT,
      region: process.env.CDK_DEFAULT_REGION,
    },
    logLevel: 'INFO',
    loginUrl: 'https://example.com/login',
    loginUserId: 'testUser',
    loginPassword: 'testPassword',
    loginBirthdate: '20000101',
    userAgent:
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3',
    line_message_api_url: 'https://api.line.me/v2/bot/message/broadcast',
    line_message_api_token: 'testLineToken',
  });
  // THEN
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
