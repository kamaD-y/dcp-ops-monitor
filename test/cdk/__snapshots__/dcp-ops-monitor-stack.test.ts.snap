// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for DcpOpsStatusNotificationStack 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-east-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs20.x",
      },
      "cn-northwest-1": {
        "value": "nodejs20.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs20.x",
      },
      "us-gov-west-1": {
        "value": "nodejs20.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-isob-west-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[Hash Id].zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "ErrorBucket8D38AF42",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ETLFunctionErrorLogAlarm03BF0B9C": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "FailureTopic74C6EA16",
          },
        ],
        "AlarmName": "ETLFunctionErrorLogAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "ETLFunctionErrorLog",
        "Namespace": "Custom/DcpOpsMonitor",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ErrorBucket8D38AF42": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "ErrorBucketAutoDeleteObjectsCustomResource8FA01807": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "ErrorBucketPolicyC8A82C62",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "ErrorBucket8D38AF42",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "ErrorBucketPolicyC8A82C62": {
      "Properties": {
        "Bucket": {
          "Ref": "ErrorBucket8D38AF42",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ErrorBucket8D38AF42",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ErrorBucket8D38AF42",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "EtlFunction22ED5A0D": {
      "DependsOn": [
        "EtlFunctionServiceRoleDefaultPolicy2FA140FD",
        "EtlFunctionServiceRole2ADAA4BC",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:[Image Tag]",
          },
        },
        "Environment": {
          "Variables": {
            "ERROR_BUCKET_NAME": {
              "Ref": "ErrorBucket8D38AF42",
            },
            "LOGIN_PARAMETER_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "LoginParametersForScrapingAA8D7353",
                  },
                ],
              ],
            },
            "LOGIN_URL": "https://example.com/login",
            "POWERTOOLS_LOG_LEVEL": "INFO",
            "SNS_TOPIC_ARN": {
              "Ref": "SuccessTopic495EEDDD",
            },
            "USER_AGENT": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
          },
        },
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "LogGroupF5B46931",
          },
        },
        "MemorySize": 1024,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "EtlFunctionServiceRole2ADAA4BC",
            "Arn",
          ],
        },
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EtlFunctionServiceRole2ADAA4BC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EtlFunctionServiceRoleDefaultPolicy2FA140FD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter",
                    {
                      "Ref": "LoginParametersForScrapingAA8D7353",
                    },
                  ],
                ],
              },
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "SuccessTopic495EEDDD",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EtlFunctionServiceRoleDefaultPolicy2FA140FD",
        "Roles": [
          {
            "Ref": "EtlFunctionServiceRole2ADAA4BC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FailureTopic74C6EA16": {
      "Type": "AWS::SNS::Topic",
    },
    "LogGroupETLFunctionErrorLogFilter613E5F93": {
      "Properties": {
        "FilterPattern": "{ $.level = "ERROR" }",
        "LogGroupName": {
          "Ref": "LogGroupF5B46931",
        },
        "MetricTransformations": [
          {
            "MetricName": "ETLFunctionErrorLog",
            "MetricNamespace": "Custom/DcpOpsMonitor",
            "MetricValue": "1",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "LogGroupF5B46931": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "LoginParametersForScrapingAA8D7353": {
      "Properties": {
        "Name": "/custom/dcp_etl/login-parameters",
        "Type": "String",
        "Value": "{"USER_ID":"testUser","PASSWORD":"testPassword","BIRTHDATE":"20000101"}",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "NotificationFunction0492F7EA": {
      "DependsOn": [
        "NotificationFunctionServiceRoleDefaultPolicy49A8A750",
        "NotificationFunctionServiceRole66D0974B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[Hash Id].zip",
        },
        "Environment": {
          "Variables": {
            "LINE_MESSAGE_API_TOKEN": "testLineToken",
            "LINE_MESSAGE_API_URL": "https://api.line.me/v2/bot/message/broadcast",
            "POWERTOOLS_LOG_LEVEL": "INFO",
          },
        },
        "Handler": "src.handler.handler",
        "Role": {
          "Fn::GetAtt": [
            "NotificationFunctionServiceRole66D0974B",
            "Arn",
          ],
        },
        "Runtime": "python3.13",
      },
      "Type": "AWS::Lambda::Function",
    },
    "NotificationFunctionAllowInvokeDcpOpsMonitorStackFailureTopic58EA38C4DFE1BC8C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "NotificationFunction0492F7EA",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "FailureTopic74C6EA16",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "NotificationFunctionAllowInvokeDcpOpsMonitorStackSuccessTopicB3D06A97EBD84C42": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "NotificationFunction0492F7EA",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "SuccessTopic495EEDDD",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "NotificationFunctionFailureTopicBF1827AA": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "NotificationFunction0492F7EA",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "FailureTopic74C6EA16",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "NotificationFunctionServiceRole66D0974B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "NotificationFunctionServiceRoleDefaultPolicy49A8A750": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "logs:DescribeMetricFilters",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:FilterLogEvents",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "LogGroupF5B46931",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "NotificationFunctionServiceRoleDefaultPolicy49A8A750",
        "Roles": [
          {
            "Ref": "NotificationFunctionServiceRole66D0974B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "NotificationFunctionSuccessTopicAAC9B438": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "NotificationFunction0492F7EA",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "SuccessTopic495EEDDD",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ScrapingNrkExecuteRule18738132": {
      "Properties": {
        "ScheduleExpression": "cron(0 0 ? * MON *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "EtlFunction22ED5A0D",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScrapingNrkExecuteRuleAllowEventRuleDcpOpsMonitorStackEtlFunction7DD83C5FE83C465A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "EtlFunction22ED5A0D",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ScrapingNrkExecuteRule18738132",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SuccessTopic495EEDDD": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
